Objects/object.c

```diff
@@ -829,7 +829,9 @@ static void
 free_object(void *obj)
 {
     PyObject *op = (PyObject *)obj;
-    Py_TYPE(op)->tp_free(op);
+    PyTypeObject *tp = Py_TYPE(op);
+    tp->tp_free(op);
+    Py_DECREF(tp);
```

Comment:
I am puzzled by this decref. Was this a refleak before? The buildbots on this pr (refleak bots) are green though

Reply:
No, there wasn't a refleak before.

Previously this code path was only used for objects with non-heap types (like float, list, dict). This PR uses it for `_asyncio._Future`, which is a heap type and needs the decref to avoid introducing a leak.

Reply:
Oh, and for static types (e.g., `PyList_Type`, `PyDict_Type`) the added `Py_DECREF` isn't a problem because the type objects are immortal.

Reply:
Right, for heaptypes decref is needed and has no effect on static types as they are immortal.

Reply:
Thanks for the clarification. For reference: the deallocation of heaptypes is described here https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_dealloc